HeterogeneousCore/SonicTriton/src/TritonService.cc

```diff
@@ -297,20 +300,39 @@ void TritonService::preBeginJob(edm::PathsAndConsumesOfModulesBase const&, edm::
                                            << output;
 }
 
+void TritonService::notifyCallStatus(bool status) const {
+  if (status)
+    --callFails_;
+  else
+    ++callFails_;
+}
+
 void TritonService::postEndJob() {
   if (!startedFallback_)
     return;
 
-  std::string command = fallbackOpts_.command + " stop";
+  std::string command = fallbackOpts_.command;
+  //prevent log cleanup during server stop
+  if (callFails_ > 0)
+    command += " -c";
+  command += " stop";
   if (verbose_)
     edm::LogInfo("TritonService") << command;
 
   const auto& [output, rv] = execSys(command);
-  if (rv != 0) {
+  if (rv != 0 or callFails_ > 0) {
+    //print logs if cmsRun is currently exiting because of a TritonException
     edm::LogError("TritonService") << output;
     printFallbackServerLog<edm::LogError>();
-    throw cms::Exception("FallbackFailed")
-        << "TritonService: Stopping the fallback server failed with exit code " << rv;
+    if (rv != 0) {
+      cms::Exception stopException("FallbackFailed");
+      stopException << "TritonService: Stopping the fallback server failed with exit code " << rv;
+      //avoid throwing if the stack is already unwinding
+      if (callFails_ > 0)
+        edm::LogWarning(stopException.category()) << stopException.explainSelf();
+      else
+        throw stopException;
+    }
```

Comment:
From the framework point of view you are technically allowed to throw an exception also in the case where the framework is shutting down because of an exception. In that case the framework has caught the exception and stored it, and strictly speaking the C++ runtime is not in a "stack unwinding" state when the framework calls `postEndJob()`. On the other hand, framework ignores all further exceptions from `endJob()`, so (at least for now) it probably makes sense to keep the printout for that case.

Thinking further about production use, we generally would not want errors in `endJob()` to lead to job failures, because the event processing itself finished. On the other hand, failing to stop the fallback server sounds a bit tricky situation, and the desired behavior might need input from WM. In the IBs I'd imagine we'd want to see these failures, so for time being I think the existing exception here is fine.

Reply:
> In that case the framework has caught the exception and stored it, and strictly speaking the C++ runtime is not in a "stack unwinding" state when the framework calls `postEndJob()`

I'm wrong here, in this case the `postEndJob()` is called as part of the stack unwinding of the first exception. But it is called within a specific `try`-`catch` block (actually nested twice), so there won't be two uncaught exceptions in flight (that would lead to immediate `std::terminate()`).

Reply:
@makortel so is the conclusion that converting from exception to warning is not needed here?